---

# Task 1: Familiarize Yourself with the Dataset

There are two main subtasks:  
1. Load the four data files that are part of this dataset.  
2. Explore each dataset and define the data types of each column. Describe the contents of each column, including the nature of the data values. Follow a similar approach as demonstrated in the lecture and tutorial. Discuss the shape of the data, value ranges, and possible categories. Convert any data into a more appropriate format if needed.  

Add as many code and markdown cells as necessary below. Typically, you should perform one type of investigation per code cell with output, and then comment on the results in a markdown cell below it.  

After completing the analysis, write a concise summary in the **Task 1: summary report** cell at the end. The final report should succinctly summarize the main characteristics of each data file, how they relate to each other, and any notable features.


### **Loading Data Files**

In [1]:
import pandas as pd

In [2]:
sessions = pd.read_csv('sessions.csv')
streamers = pd.read_csv('streamers.csv')
streams = pd.read_csv('streams.csv')
viewers = pd.read_csv('viewers.csv')

### **Dataset Exploration**

**Sessions**

In [3]:
sessions.head()

,session_id,viewer_id,streamer_id,stream_id,started_at,ended_at,duration_mins,chat_messages_sent,bits_cheered,followed_during,subscribed_during
0,SES000001,VWR00938,STR0055,SM00860,2024-01-18 19:16:00,2024-01-18 20:28:00,72,0,0,False,False
1,SES000002,VWR00804,STR0034,SM00542,2024-01-04 17:11:00,2024-01-04 19:41:00,150,11,0,False,False
2,SES000003,VWR00704,STR0008,SM00124,2024-02-15 15:11:00,2024-02-15 16:53:00,102,5,0,False,False
3,SES000004,VWR00926,STR0024,SM00373,2024-01-18 19:59:00,2024-01-18 20:56:00,57,3,9663,False,False
4,SES000005,VWR00883,STR0008,SM00120,2024-03-19 20:08:00,2024-03-19 20:22:00,14,7,0,False,False


Looking at this data file, I can see that each session, viewer, streamer, and stream has its own ID composed of letters and numbers. There are timestamps for the start and end of the stream. Numeric values are used to express the duration in munutes, number of chats sent, and number of cheers (chats sent with bits, virtual currency). I can see that booleans (True or False) are used to tell whether or not people followed during the stream and if they subscribed during the stream.

In [4]:
sessions.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 22825 entries, 0 to 22824
Data columns (total 11 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   session_id          22825 non-null  object
 1   viewer_id           22825 non-null  object
 2   streamer_id         22765 non-null  object
 3   stream_id           22825 non-null  object
 4   started_at          22825 non-null  object
 5   ended_at            22825 non-null  object
 6   duration_mins       22825 non-null  int64 
 7   chat_messages_sent  22825 non-null  int64 
 8   bits_cheered        22825 non-null  int64 
 9   followed_during     22825 non-null  bool  
 10  subscribed_during   22825 non-null  bool  
dtypes: bool(2), int64(3), object(6)
memory usage: 1.6+ MB


There are a total of 22825 entries with a total of 11 columns. There are null values only in the streamer_id column. The values are a mix of data types, including objects (usually strings), ints, and booleans.

In [5]:
sessions.describe()

,duration_mins,chat_messages_sent,bits_cheered
count,22825.000000,22825.000000,22825.000000
mean,59.051303,6.442629,398.425323
std,60.383979,46.576309,1571.528658
min,0.000000,0.000000,0.000000
25%,15.000000,0.000000,0.000000
50%,39.000000,1.000000,0.000000
75%,83.000000,5.000000,0.000000
max,443.000000,1987.000000,10000.000000


For the numeric data in this file, streams average about 59 minutes with a minimum length of 0 minutes and maximum length of 443 minutes. The average amount of chat messages sent is 6 with a minimum of 0 and maximum of 1987. For cheers, there is an average of 398 with a minimum of 0 and maximum of 10,000. Looking at the entire dataset, it seems like more people are commenting but in smaller amounts of comments. For cheers, there are fewer instances of people sending bits, but when they do, they send a large amount of bits.

**Streamers**

In [6]:
streamers.head()

,streamer_id,streamer_name,category,language,partner_status,total_followers,avg_concurrent_viewers
0,STR0001,xQc,Just Chatting,English,True,11500000,138706
1,STR0002,Pokimane,Just Chatting,English,True,-25551,71660
2,STR0003,HasanAbi,Just Chatting,English,True,2600000,34220
3,STR0004,Amouranth,Just Chatting,English,True,6200000,68607
4,STR0005,Mizkif,Just Chatting,English,True,2100000,6771


This data file focuses on each streamer, who gets a unique ID. The data includes the streamer's name, their streaming category, language, whether or not the streamer is a partner, how many followers they have, and their average number of concurrent viewers. Here there is also a combination of string, boolean, and numeric data. The streamer's category and language is categorical data.

In [7]:
streamers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 76 entries, 0 to 75
Data columns (total 7 columns):
 #   Column                  Non-Null Count  Dtype 
---  ------                  --------------  ----- 
 0   streamer_id             76 non-null     object
 1   streamer_name           76 non-null     object
 2   category                76 non-null     object
 3   language                68 non-null     object
 4   partner_status          76 non-null     bool  
 5   total_followers         76 non-null     int64 
 6   avg_concurrent_viewers  76 non-null     int64 
dtypes: bool(1), int64(2), object(4)
memory usage: 3.8+ KB


There are a total of 76 rows of data across 7 columns. There is data in the form of objects (strings), booleans, and integers. One column, language, contains 8 null values.

In [8]:
streamers.describe().round(2)

,total_followers,avg_concurrent_viewers
count,76.00,76.00
mean,3300885.76,31227.99
std,3793772.48,38307.22
min,-46964.00,634.00
25%,780000.00,6633.25
50%,2000000.00,14762.50
75%,3875000.00,36840.75
max,18700000.00,156401.00


The average amount of followers for this group of streamers is 3,300,886 and the average amount of concurrent viewers is 31,227. It looks like the minimum amount of followers is -46964, and I assume that negative numbers in this column are data entry issues meant to be positive since it's impossible to have a negative following count. The most followers any streamer in this list has is 18,700,000. The maximum amout of average viewers by a streamer is 156,401 and the minimum average viewers is 634.

**Streams**

In [9]:
streams.head()

,stream_id,streamer_id,started_at,ended_at,stream_duration_hrs,category,peak_viewers,title_has_hype_word,was_raid
0,SM00001,STR0001,2024-02-14 18:00:00,2024-02-14 21:00:00,3.0,Just Chatting,108989.0,False,True
1,SM00002,STR0001,2024-01-28 18:00:00,2024-01-28 21:00:00,3.0,Just Chatting,51635.0,True,False
2,SM00003,STR0001,2024-01-31 09:00:00,2024-01-31 16:00:00,7.0,Just Chatting,212127.0,False,False
3,SM00004,STR0001,2024-03-03 14:00:00,2024-03-03 21:00:00,7.0,Just Chatting,186745.0,False,True
4,SM00005,STR0001,2024-01-24 18:00:00,2024-01-24 21:00:00,3.0,Just Chatting,87723.0,False,True


This data file contains data about each individual streaming session. Each stream and streamer has a unique ID which is a combination of letters and numbers. There are timestamps for start and end time of the stream. The stream's duration is numerical data in hours. The stream's category is categorical data. The maximum number of viewers during the duration of the stream is peak_viewers and is expressed as numerical data. Boolean data, True or False, describes whether or not there were hype words in the stream and if there was a raid during the stream.

In [10]:
streams.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1199 entries, 0 to 1198
Data columns (total 9 columns):
 #   Column               Non-Null Count  Dtype  
---  ------               --------------  -----  
 0   stream_id            1199 non-null   object 
 1   streamer_id          1199 non-null   object 
 2   started_at           1199 non-null   object 
 3   ended_at             1199 non-null   object 
 4   stream_duration_hrs  1199 non-null   float64
 5   category             1149 non-null   object 
 6   peak_viewers         1119 non-null   float64
 7   title_has_hype_word  1199 non-null   bool   
 8   was_raid             1199 non-null   bool   
dtypes: bool(2), float64(2), object(5)
memory usage: 68.0+ KB


This data file contains data in the form of objects, floats, and booleans. There is a total of 1199 rows with 9 columns. The category and peak_viewers columns are missing values.

In [11]:
streams.describe().round(2)

,stream_duration_hrs,peak_viewers
count,1199.00,1119.00
mean,4.63,38375.66
std,8.00,50377.09
min,1.50,-50.00
25%,2.50,7598.50
50%,3.00,20224.00
75%,5.00,47941.00
max,96.00,370165.00


Streams are about 4.6 hours on average with the minimum length being 1.5 hours and the maximum 96 hours. Peak viewers is missing 80 values. Number of average viewers is 38,376 viewers with the minimum being -50 and maximum being 370,165. It doesn't make sense to have a negative number of viewers, so the minimum value and anything less than 0 must be a data input error.

**Viewers**

In [12]:
viewers.head()

,viewer_id,age_group,country,account_age_days,subscription_tier,preferred_category
0,VWR00001,18-24,AR,355,free,IRL
1,VWR00002,25-34,CA,1670,free,Creative
2,VWR00003,35-44,US,1108,free,Just Chatting
3,VWR00004,45+,NaN,822,tier1,FPS
4,VWR00005,25-34,DE,2226,free,Just Chatting


This data file is all about the viewers. Each viewer has their own unique ID which is a combination of letters and numbers. They also have a specified age group which is categorical. Each viewer also has a country, also categorical. Their account age is in days and is numerical data. Their subscription tier, either free, tier 1, tier 2, tier 3, or premium, is categorical, as is their preferred category.

In [13]:
viewers.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 6 columns):
 #   Column              Non-Null Count  Dtype 
---  ------              --------------  ----- 
 0   viewer_id           1000 non-null   object
 1   age_group           1000 non-null   object
 2   country             920 non-null    object
 3   account_age_days    1000 non-null   int64 
 4   subscription_tier   1000 non-null   object
 5   preferred_category  1000 non-null   object
dtypes: int64(1), object(5)
memory usage: 47.0+ KB


This data file contains data in the form of objects and integers. There are a total of 1000 rows with 6 columns. The country column has 80 missing values.

In [14]:
viewers.describe()

,account_age_days
count,1000.000000
mean,1881.119000
std,1036.144656
min,33.000000
25%,984.500000
50%,1966.500000
75%,2761.250000
max,3645.000000


The average account is 1881.1 days old, with the youngest account at 33 days old and the oldest account at 3645 days old. The minimum and maximum look correct here. I had to check the age of Twitch, as the oldest account is just over 10 years old. This seems possible as Twitch launched in 2011.

### Task 1 Summary Report

Write your report here, summarizing the work done in the cells above.

# Task 2: Data Cleaning, Diagnostics, and Corrections

Identify and resolve issues in the dataset. These can be of different kinds, such as:
- Look for duplicates or implausible values (e.g., users with negative followers, invalid age groups). Thoroughly document your exploration of the data. Decide whether these are outliers or the result of a systemic problem. If the latter, resolve the problem here; if they are outliers, we will address those later.  
- Define at least three heuristics (common-sense rules) to flag unusual entries (e.g., someone watching dozens of streams per day). Discuss whether what you found are errors or artifacts.  

I will repeat the instructions here, but this is expected for each task: each code cell should focus on a specific analysis. You should run the code, and in the markdown below each code cell and its result, provide a detailed interpretation of what you have uncovered in that step. After completing the analyses, summarize the key findings in the indicated summary report cell below.


### **Duplicates or Implausible Values**

**Negative Values**

From Task 1, I noticed negative values for total followers in the streamers data and peak viewers in the streams data. Assuming these are data entry issues, since these values cannot possibly be less than 0, I will simply turn them positive.

In [15]:
#Streamers data

#Turn all negative values positive
streamers['total_followers'] = streamers['total_followers'].abs()
#Let's see if the new minimum is positive
streamers['total_followers'].describe().round(2)

,total_followers
count,76.00
mean,3305166.87
std,3789993.58
min,8902.00
25%,780000.00
50%,2000000.00
75%,3875000.00
max,18700000.00


In [16]:
#Streams data

#Turn all negative values positive
streams['peak_viewers'] = streams['peak_viewers'].abs()

#Let's see if the new minimum is positive
streams['peak_viewers'].describe().round(2)

,peak_viewers
count,1119.00
mean,38377.59
std,50375.61
min,1.00
25%,7598.50
50%,20224.00
75%,47941.00
max,370165.00


**Duplicate Values**

In [17]:
#Print the number of duplicate rows in each data file

print(sessions.duplicated().sum())
print(streamers.duplicated().sum())
print(streams.duplicated().sum())
print(viewers.duplicated().sum())

0
0
0
0


There appear to be no duplicate rows in the dataset.

In [18]:
print(sessions['session_id'].duplicated().sum())
print(sessions['viewer_id'].duplicated().sum())
print(sessions['streamer_id'].duplicated().sum())
print(sessions['stream_id'].duplicated().sum())

0
21825
22748
21591


It looks like there are a lot of duplicate viewer IDs, streamer IDs, and stream IDs. It makes sense in this case as there are multiple data entries on the same viewers, streamers, and streams.

In [19]:
print(streamers['streamer_id'].duplicated().sum())
print(streamers['streamer_name'].duplicated().sum())

0
0


There are no duplicate streamers in the streamers data file.

In [20]:
print(streams['stream_id'].duplicated().sum())

0


There are no duplicate stream IDs in the streams data file.

In [21]:
print(viewers['viewer_id'].duplicated().sum())

0


There are no duplicate viewer IDs in the viewers file.

**Checking Timestamps**

In [22]:
#Convert timestamps in sessions data to datetimes
sessions['started_at'] = pd.to_datetime(sessions['started_at'])
sessions['ended_at'] = pd.to_datetime(sessions['ended_at'])

#Convert timestamps in streams data to datetimes
streams['started_at'] = pd.to_datetime(streams['started_at'])
streams['ended_at'] = pd.to_datetime(streams['ended_at'])

Looking at our class discussion and at the data files in Google Sheets, some of the timestamps appear to be swapped. This means that the start time is after the end time, and this doesn't make logical sense.

In [23]:
checkReversed = sessions['started_at'] > sessions['ended_at']
print(checkReversed.sum())


40


There are 40 rows with reversed start times and end times. I also want to find out if the computed number of minutes matches the logged minutes.

In [24]:
#Compute number of minutes from the start time and end time
computedMins = (sessions['ended_at'] - sessions['started_at']).dt.total_seconds() / 60

#Check if the computed minutes match the logged duration_mins
matches = ((computedMins[checkReversed].abs() - sessions.loc[checkReversed, 'duration_mins']).abs() < 2).all()
print(matches)

True


It is true that the computed minutes match the logged minutes. Hooray.

In [25]:
sessions.loc[checkReversed, ['session_id', 'started_at', 'ended_at', 'duration_mins']].head()

,session_id,started_at,ended_at,duration_mins
341,SES000342,2024-03-19 13:53:00,2024-03-19 12:15:00,98
722,SES000723,2024-03-15 14:53:00,2024-03-15 11:57:00,176
1011,SES001012,2024-02-19 08:36:00,2024-02-19 08:18:00,18
1718,SES001719,2024-02-13 19:38:00,2024-02-13 19:36:00,2
1897,SES001898,2024-01-01 11:18:00,2024-01-01 09:14:00,124


Quick check to make sure this looks okay. Now time to fix the issue.

In [26]:
#For reversed rows, swap the started_at and ended_at values
sessions.loc[checkReversed, ['started_at', 'ended_at']] = (
    sessions.loc[checkReversed, ['ended_at', 'started_at']].values
)

#Check that this worked
remaining = (sessions['ended_at'] < sessions['started_at']).sum()
print(remaining)

0


There are 0 rows left where the end time is before the start time.

**Logged Duration vs Computed Duration**

In streams.csv, there were a few values that raised suspicion, like when I saw the 96 hour maximum duration. That's a long stream. For example, SM00547 has a logged duration of 96 hours, but the start time is 12:00 and the end time is 16:00 on the same day. The logged duration should only be 4 hours.

In [27]:
#Compute number of hours using start and end time
streams['computedHours'] = ((streams['ended_at'] - streams['started_at']).dt.total_seconds() / 3600)

#Compare computed duration with logged duration
wrongDuration = (streams['computedHours'] - streams['stream_duration_hrs']).abs() > 0.1
print(wrongDuration.sum())

15


There are 15 streams in which the logged hours don't match the computed streaming time.

In [28]:
#Rows with wrong durations, showing logged duration vs computed duration
streams.loc[wrongDuration, ['stream_id', 'streamer_id', 'started_at', 'ended_at', 'stream_duration_hrs', 'computedHours']]

,stream_id,streamer_id,started_at,ended_at,stream_duration_hrs,computedHours
6,SM00007,STR0001,2024-03-05 12:00:00,2024-03-05 16:00:00,48.0,4.0
48,SM00049,STR0003,2024-02-02 10:00:00,2024-02-02 15:00:00,72.0,5.0
284,SM00285,STR0018,2024-01-24 14:00:00,2024-01-24 17:00:00,72.0,3.0
332,SM00333,STR0021,2024-02-12 22:00:00,2024-02-13 02:00:00,72.0,4.0
357,SM00358,STR0023,2024-03-16 09:00:00,2024-03-16 12:00:00,48.0,3.0
364,SM00365,STR0024,2024-02-28 22:00:00,2024-02-29 02:00:00,72.0,4.0
372,SM00373,STR0024,2024-01-18 15:00:00,2024-01-18 22:00:00,48.0,7.0
417,SM00418,STR0026,2024-03-15 21:00:00,2024-03-16 02:00:00,72.0,5.0
458,SM00459,STR0029,2024-02-25 21:00:00,2024-02-25 23:30:00,48.0,2.5
546,SM00547,STR0034,2024-02-25 12:00:00,2024-02-25 16:00:00,96.0,4.0


In [29]:
#Replace logged duration with computed duration

streams.loc[wrongDuration, 'stream_duration_hrs'] = (streams.loc[wrongDuration, 'computedHours'])

#Check that it worked and drop the computedHours column that I made
streams.drop(columns='computedHours')
streams.loc[wrongDuration, ['stream_id', 'streamer_id', 'started_at', 'ended_at', 'stream_duration_hrs']]

,stream_id,streamer_id,started_at,ended_at,stream_duration_hrs
6,SM00007,STR0001,2024-03-05 12:00:00,2024-03-05 16:00:00,4.0
48,SM00049,STR0003,2024-02-02 10:00:00,2024-02-02 15:00:00,5.0
284,SM00285,STR0018,2024-01-24 14:00:00,2024-01-24 17:00:00,3.0
332,SM00333,STR0021,2024-02-12 22:00:00,2024-02-13 02:00:00,4.0
357,SM00358,STR0023,2024-03-16 09:00:00,2024-03-16 12:00:00,3.0
364,SM00365,STR0024,2024-02-28 22:00:00,2024-02-29 02:00:00,4.0
372,SM00373,STR0024,2024-01-18 15:00:00,2024-01-18 22:00:00,7.0
417,SM00418,STR0026,2024-03-15 21:00:00,2024-03-16 02:00:00,5.0
458,SM00459,STR0029,2024-02-25 21:00:00,2024-02-25 23:30:00,2.5
546,SM00547,STR0034,2024-02-25 12:00:00,2024-02-25 16:00:00,4.0


The stream_duration_hrs that did not match were replaced with the computed time using the start and end timestamps.

For entries that didn't have a reversed start time and end time, I also need to check if computed duration matches logged duration.

In [30]:
#Checking for differences between computed and logged
sessions['computedMins'] = ((sessions['ended_at'] - sessions['started_at']).dt.total_seconds() / 60)
sessions['durationDiff'] = (sessions['computedMins'] - sessions ['duration_mins']).abs()

mismatched = ((sessions['durationDiff'] > 1) & (sessions['computedMins'] >0))

print(mismatched.sum())

30


There are 30 entries in which the computed duration and logged duration differ by over a minute.

In [31]:
#Replacing logged duration with computed duration
sessions.loc[mismatched, 'duration_mins'] = (sessions.loc[mismatched, 'computedMins'].round().astype(int))

The mismatched logged minutes have been replaced with computed minutes using the start and end timestamps.

**Subscription Tiers**

As found by me and Johann for the group bonus challenge, there are errors with data entry in subscription_tier in viewers.csv. Tier 1 viewers are entered as both "tier1" and "TIER1". Viewers with the free tier have been entered as "none", "free trial", and "free". As well, Twitch does not have a tier 4 or premium tier, which have been entered into this dataset.

In [32]:
print(viewers['subscription_tier'].unique())

['free' 'tier1' 'tier2' 'tier3' 'free trial' 'premium' 'none' 'tier4'
 'TIER1']


Here are all the different entries for the subscription_tier column. There should really only be free, tier1, tier2, and tier3.

In [33]:
#Map messed up values to good values
tierMap = {
    'free trial' : 'free',
    'none' : 'free',
    'TIER1' : 'tier1',
    'tier4' : 'tier3',
    'premium' : 'tier3'
}

#Replace values using the map
viewers['subscription_tier'] = viewers['subscription_tier'].replace(tierMap)

#Check if this worked
print(viewers['subscription_tier'].unique())

['free' 'tier1' 'tier2' 'tier3']


There are now only 4 categories for tiers and the naming is consistent.

**Age Groups**

In [34]:
print(viewers['age_group'].unique())

['18-24' '25-34' '35-44' '45+' '13-17' '18–24' '25–34' '35 to 44'
 '18 to 24' '25 to 34' '35–44']


As mentioned by classmates in Discord, there are inconsistent naming conventions for the age groups. Some use different types of dashes, other use "to".

In [35]:
#Map messed up age values to good age values
ageMap = {
    '18\u201324' : '18-24',
    '18 to 24' : '18-24',
    '25\u201334' : '25-34',
    '25 to 34' : '25-34',
    '35\u201344' : '35-44',
    '35 to 44' : '35-44'
}

#Replace values using the map
viewers['age_group'] = viewers['age_group'].replace(ageMap)

#Check if this worked
print(viewers['age_group'].unique())

['18-24' '25-34' '35-44' '45+' '13-17']


There are now 5 specified age groups with the same naming conventions.

**Heuristics**

*Chat Messages*

When looking at sessions.csv I noticed that the maximum value for the chat_messages column was 1987. I saw this when doing my initial exploration above in this notebook. That's a lot of messages for a person to send in a single session, and when I looked at the entry, the session in which 1987 messages were sent was only 4 minutes long. This is physically impossible. I'm flagging this (and other entries where there are >20 messages per minute logged) as an unusual entry and as potential bot activity.

This would be an artifact, not an error. It's possible that these events really happened, especially if the user was a bot. However, these entries with >20 chat messages/minute are not representative of the behaviour of human viewers and might have to be excluded from analysis if the goal is observing engagement.

*0 Mins with Chats*

As shown below, there appear to be some entries in sessions.csv with 0 logged duration minutes but a number of chats. It's pretty much impossible to send chats if the session lasted 0 minutes.

This is probably an error, especially the entries with a high number of chats but 0 minutes of duration. The timestamp was either recorded at the wrong resolution (like in seconds instead of minutes), or the chat activity is incorrect/in the wrong record. This is unrecoverable information.

In [39]:
#Some of these sessions with 0 minutes recorded still have chat activity
sessions.loc[sessions['duration_mins'] == 0]

,session_id,viewer_id,streamer_id,stream_id,started_at,ended_at,duration_mins,chat_messages_sent,bits_cheered,followed_during,subscribed_during,computedMins,durationDiff
110,SES000111,VWR00984,STR0009,SM00143,2024-02-07 13:39:00,2024-02-07 13:39:00,0,2,0,False,False,0.0,0.0
306,SES000307,VWR00520,STR0019,SM00288,2024-01-13 19:45:00,2024-01-13 19:45:00,0,0,4406,False,False,0.0,0.0
416,SES000417,VWR00347,STR0041,SM00659,2024-03-23 15:40:00,2024-03-23 15:40:00,0,1,0,False,False,0.0,0.0
826,SES000827,VWR00771,STR0044,SM00692,2024-02-25 18:50:00,2024-02-25 18:50:00,0,1,0,False,False,0.0,0.0
899,SES000900,VWR00247,STR0061,SM00952,2024-03-14 09:33:00,2024-03-14 09:33:00,0,0,6404,False,False,0.0,0.0
1029,SES001030,VWR00676,STR0056,SM00877,2024-01-06 22:01:00,2024-01-06 22:01:00,0,0,0,False,False,0.0,0.0
1325,SES001326,VWR00716,STR0005,SM00079,2024-03-23 21:13:00,2024-03-23 21:13:00,0,9,0,False,False,0.0,0.0
2046,SES002047,VWR00901,STR0026,SM00420,2024-02-14 16:50:00,2024-02-14 16:50:00,0,3,0,False,False,0.0,0.0
3009,SES003010,VWR00496,STR0072,SM01137,2024-02-20 20:47:00,2024-02-20 20:47:00,0,1,6578,False,False,0.0,0.0
3459,SES003460,VWR00666,STR0038,SM00617,2024-01-12 20:11:00,2024-01-12 20:11:00,0,1,0,False,False,0.0,0.0


*Overlapping Sessions*

Looking at timestamps again, there are some overlapping started_at and ended_at times for the same viewer_id on the same day. This shouldn't be possible since I think Twitch only allows a user to watch one stream at a time.

This is likely an artifact, an issue with the measurement/logging method. The end time may have been accidentally logged as the time the stream ended, not when the viewer left. In this case, the duration might be more reliable than the timestamps and will have to be reviewed.

### Task 2 summary report

# Task 3: Descriptive Statistics and Visualizations

- Create summary statistics and visualizations to understand user characteristics  
- Create summary statistics for streams data overall, and summaries by age group, account tire, or peak viewers  
- Visualize the distribution of streams durations, popular categories,   
- Show category-specific information and compare streamers based on categories  

In [ ]:
import matplotlib.pyplot as plt
import seaborn as sns

### Task 3 Summary report
summarize your findings here.

# Task 4: Dealing with Outliers and Missing Values

- Are there any missing values or outliers in the original data? Perform the analysis and report your findings.  
- Are there any outliers or missing values in the derived data? Perform the analysis and report your findings.  
- If there are no outliers or missing values, introduce 3–5% of missing values in a selected column and 1–2% of outliers in another column.  
- Experiment with two strategies for imputing missing values and two strategies for handling outliers. Show how each strategy affects the characteristics of the data.


In [ ]:
#start code for Task 4 here

### Task 4 Summary report

# Task 5: Correlation and Behavioral Patterns

- Compute pairwise correlations between user features and the number of check-ins  
- Visualize the results using a correlation matrix (use a coolwarm color scheme and proper scaling)  
- Try Spearman rank correlation and explain when it is more appropriate than Pearson

In [ ]:
# tast 5 code starts here

### Task 5 Summary report

# Task 6: Compare Behavior Across User Segments

- Segment users into 2–3 groups (e.g., by age bracket, account tire)  
- Compare their viewing behavior over time using boxplots or dot plots  
- Identify any outliers in group behavior and investigate them

In [ ]:
# task 6 code start here

### Task 6 Summary report

# Task 7: Your Own EDA Questions

Formulate and explore two hypotheses of your own:  
- Propose questions based on curiosity (e.g., “Does a user's variety of categories influence morning vs. evening or session duration?”)  
- Use summary statistics and visualizations to answer each question  
- Briefly discuss what additional data would help you answer the questions more effectively


In [ ]:
# task 7 code start here

### Task 7 Summary report

# Task 8: EDA Presentation (Insight Report)

Summarize your findings in a polished format:  
- Create 3–4 slides or a single-page report highlighting your most important insights  
- Include at least 2 charts and interpret them using clear, plain language  
- Apply Tufte’s principles: minimize chartjunk, maximize data-to-ink ratio, and ensure clear labeling  